File: implementation-go/wallet-service/handler.go
package wallet_service

import (
	"context"
	"go-microtrans-grpc/wallet-service/pb/wallet"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

type WalletHandler struct {
	wallet.UnimplementedWalletServiceServer
	service *WalletService
}

func (h *WalletHandler) ProcessTransaction(ctx context.Context, req *wallet.TransactionRequest) (*wallet.TransactionResponse, error) {
	userId := req.GetUserId()
	amount := req.GetAmount()
	transactionType := req.GetTransactionType()

	if userId <= 0 {
		return nil, status.Error(codes.InvalidArgument, "User ID invalid")
	}

	if amount <= 0 {
		return nil, status.Error(codes.InvalidArgument, "Amount must be positive")
	}

	if TransactionType(transactionType) == Unknown {
		return nil, status.Error(codes.InvalidArgument, "Transaction type invalid")
	}

	return h.service.ExecuteTransaction(ctx, req)
}

func NewWalletHandler(service *WalletService) *WalletHandler {
	return &WalletHandler{
		service: service,
	}
}

File: implementation-nodejs/wallet-service/src/handler.js
const grpc = require('@grpc/grpc-js');

class WalletHandler 
{
    constructor(service)
    {
        this.service = service;
    }

    async processTransaction(call, callback) 
    {
        try 
        {
            const { user_id, amount, transaction_type } = call.request;

            const result = await this.service.executeTransaction(user_id, amount, transaction_type);

            callback(null, {
                success: result.success,
                message: result.message,
                transaction_id: result.transaction_id,
                final_balance: result.final_balance
            });
        } 
        catch (error) 
        {
            callback({
                code: grpc.status.INTERNAL,
                details: error.message
            });
        }
    }
}

module.exports = WalletHandler;